histogram/src/histogram.rs

```diff
@@ -0,0 +1,430 @@
+        self
+    }
+}
+
+impl Histogram {
+    #[allow(non_snake_case)]
+    pub fn new(m: u32, r: u32, n: u32) -> Self {
+        if r <= m || r > n || n > 64 {
+            panic!("invalid histogram config");
+        }
+
+        let M = 1 << m;
+        let R = (1 << r) - 1;
+        let N = (1 << n) - 1;
+        let G: u64 = 1 << (r - m - 1);
+
+        let n_buckets = (n - r + 2) as u64 * G;
+
+        let mut buckets = Vec::new();
+        buckets.resize_with(n_buckets as usize, || AtomicU32::new(0));
+
+        Self {
+            m,
+            r,
+            n,
+            M,
+            R,
+            N,
+            G,
+            buckets: buckets.into_boxed_slice(),
+        }
+    }
+
+    /// Creates a `Builder` with the default values `m = 0`, `r = 10`, `n = 30`.
+    ///
+    /// This would create a `Histogram` with 11264 buckets which can store
+    /// values from 1 to 1_073_741_823 with values 1 to 1023 being stored in
+    /// buckets with a width of 1. Such a `Histogram` would be appropriate for
+    /// latencies measured in nanoseconds where the max expected latency is one
+    /// second.
+    pub fn builder() -> Builder {
+        Builder { m: 0, r: 10, n: 30 }
+    }
+
+    /// Resets the `Histogram` by zeroing out the count for every bucket.
+    pub fn clear(&self) {
+        for bucket in self.buckets.iter() {
+            bucket.store(0, Ordering::Relaxed);
+        }
+    }
+
+    /// Increment the histogram bucket corresponding to the provided `value` by
+    /// the provided `count`.
+    ///
+    /// This operation wraps on overflow.
+    #[allow(clippy::result_unit_err)]
+    pub fn increment(&self, value: u64, count: u32) -> Result<(), Error> {
+        if value > self.N {
+            // value too big
+            return Err(Error::OutOfRange);
+        }
+
+        let index = self.bucket_index(value);
+        self.buckets[index].fetch_add(count, Ordering::Relaxed);
+
+        Ok(())
+    }
+
+    /// Decrement the histogram bucket corresponding to the provided `value` by
+    /// the provided `count`.
+    ///
+    /// This operation wraps on overflow.
+    #[allow(clippy::result_unit_err)]
+    pub fn decrement(&self, value: u64, count: u32) -> Result<(), Error> {
+        if value > self.N {
+            // value too big
+            return Err(Error::OutOfRange);
+        }
+
+        let index = self.bucket_index(value);
+        self.buckets[index].fetch_add(count, Ordering::Relaxed);
+
+        Ok(())
+    }
+
+    /// Retrieve the `Bucket` which corresponds to the provided percentile.
+    ///
+    /// An error will be returned if the percentile is invalid or if there are
+    /// no samples in the `Histogram`.
+    ///
+    /// Note: if you are reporting on multiple percentiles, it is more efficient
+    /// to use the `percentiles` function to retrieve multiple percentiles in a
+    /// single call.
+    #[allow(clippy::result_unit_err)]
```

Comment:
This isn't returning `Result<(), Error>`. Does it need to suppress the warning?

Reply:
Fixed.